File: src/main/java/weld/WeldManaged.java
package weld;

/**
 * Base class for classes for which an object exists that is managed by weld. All these classes
 * must be explicitly closed after use in order to prevent resource leaks.
 */
public abstract class WeldManaged implements AutoCloseable {
  /**
   * Handle to the native object.
   */
  final long handle;

  /**
   * Flag to indicate that the value is closed.
   */
  private boolean closed = false;

  /**
   * Create a new weld managed object.
   *
   * @param handle to the native object.
   */
  WeldManaged(long handle) {
    super();
    this.handle = handle;
  }

  /**
   * Close the weld managed object. Note that this method is idempotent.
   */
  @Override
  public final synchronized void close() {
    if (!closed) {
      doClose();
      closed = true;
    }
  }

  /**
   * Close the weld managed object.
   */
  protected abstract void doClose();

  /**
   * Check if the weld managed is closed.
   */
  public boolean closed() {
    return closed;
  }

  void checkAccess() {
    assert !closed : "Cannot access an already closed object";
  }
}

File: src/main/scala/weld/WeldManaged.scala
package weld

/**
 * Base class for classes for which an object exists that is managed by weld. All these classes
 * must be explicitly closed after use in order to prevent resource leaks.
 */
abstract class WeldManaged(val handle: Long) extends AutoCloseable {
  /**
   * Flag to indicate that the value is closed.
   */
  private var closed = false

  /**
   * Flag to indicate that the underlying reference will be cleaned up as soon as this object is
   * garbage collected.
   */
  private var autoClean = false

  /**
   * Close the weld managed object. Note that this method is idempotent.
   */
  override def close(): Unit = {
    if (!closed && !autoClean) {
      doClose()
      closed = true
    }
  }

  /**
   * Sets the closeable to auto cleaning mode. This means that the underlying handle is cleaned up
   * as soon as the object is garbage collected.
   */
  def markAutoCleanable(): Unit = {
    if (!closed && !autoClean) {
      autoClean = true
      Platform.registerForCleanUp(this, cleaner)
    }
  }

  /**
   * Close the weld managed object.
   */
  protected def doClose(): Unit

  /**
   * Create a cleaner for the managed object.
   */
  protected def cleaner: Runnable

  /**
   * Check if the weld managed is closed.
   */
  def isClosed(): Boolean = closed

  private[weld] def checkAccess() = {
    assert(!closed, "Cannot access an already closed object")
  }
}
